overlord/exportstate/manifest.go

```diff
@@ -0,0 +1,209 @@
+	switch info.Type() {
+	case snap.TypeSnapd:
+		return manifestForSnapdSnap(info)
+	case snap.TypeOS: // we really mean "core" snap here.
+		return manifestForCoreSnap(info)
+	default:
+		// TODO: add support for regular snaps
+		return nil
+	}
+}
+
+// NewManifestForHost returns the export manifest of the host system.
+//
+// Classic systems have an export manifest which points to the snapd
+// tools from the classic (deb/rpm) packages. Ubuntu Core systems do
+// not need this, the snapd tools come from the snapd snap only.
+func NewManifestForHost() *Manifest {
+	if release.OnClassic {
+		return manifestForClassicSystem()
+	}
+	// XXX: should this just return nil instead? the core system manifest
+	// is empty
+	return manifestForCoreSystem()
+}
+
+// createExportedFiles creates all the files constituting the export manifest.
+//
+// The directory /var/lib/snapd/export/$source/$exportedVersion is created
+// if necessary. For each export set in the manifest, additional sub-directory
+// is created and populated with symbolic links pointing to the exported files.
+//
+// The function is idempotent.
+func createExportedFiles(manifest *Manifest) error {
+	for _, set := range manifest.Sets {
+		for _, exported := range set.Exports {
+			if err := createExportedFile(manifest, &set, &exported); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// removeExportedFiles removes all the files constituting the export state.
+//
+// In addition the path /var/lib/snapd/export/$source/$exportedVersion
+// is pruned, removing empty directories if possible.
+//
+// On failure removal continues and the first error is returned.
+func removeExportedFiles(manifest *Manifest) error {
+	var firstErr error
+	for _, set := range manifest.Sets {
+		for _, exported := range set.Exports {
+			if err := removeExportedFile(manifest, &set, &exported); err != nil && firstErr == nil {
+				firstErr = err
+			}
+		}
+	}
+	return firstErr
+}
+
+// exportedFilePath returns the path of an exported file.
+func exportedFilePath(manifest *Manifest, set *ExportSet, exported *ExportedFile) string {
+	snapInstanceName := manifest.SnapInstanceName
+	snapRevision := manifest.SnapRevision.String()
+	if manifest.ExportedForSnapdAsVersion != "" { // Exception for core and host
+		snapInstanceName = "snapd"
+		snapRevision = manifest.ExportedForSnapdAsVersion
+	}
+	return filepath.Join(dirs.ExportDir, snapInstanceName, snapRevision, set.Name, exported.Name)
+}
+
+// exportedFileSourcePath returns the source path that is to be exported.
+func exportedFileSourcePath(manifest *Manifest, set *ExportSet, exported *ExportedFile) string {
+	// Consumer uses host mount namespace
+	if set.ConsumerIsHost {
+		if manifest.SourceIsHost {
+			// host-to-host, no translation needed.
+			return exported.SourcePath
+		}
+		// snap-to-host
+		return filepath.Join(dirs.SnapMountDir, manifest.SnapInstanceName, manifest.SnapRevision.String(), exported.SourcePath)
```

Comment:
I probably need a picture to visualize this :-) , but don't we need to do the same check on `manifest.ExportedForSnapdAsVersion` here (and in the last line of this function)?